src/main/java/org/jetbrains/kotlinx/lincheck/strategy/managed/TraceReporter.kt

```diff
@@ -0,0 +1,271 @@
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.jetbrains.kotlinx.lincheck.strategy.managed
+
+import org.jetbrains.kotlinx.lincheck.*
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionResult
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario
+import org.jetbrains.kotlinx.lincheck.execution.parallelResults
+import org.jetbrains.kotlinx.lincheck.printInColumnsCustom
+import kotlin.math.min
+
+private const val INTERLEAVING_INDENTATION = "  "
+
+internal fun StringBuilder.appendTrace(
+    scenario: ExecutionScenario,
+    results: ExecutionResult?,
+    trace: List<InterleavingEvent>
+) {
+    // clear object numeration that is used by `CodePoint.toString` for better representation
+    objectNumeration.clear()
+    val graphStart = constructInterleavingGraph(scenario, results, trace)
+    val interleaving = interleavingGraphToRepresentation(graphStart)
+    val interleavingRepresentation = splitToColumns(scenario.threads, interleaving)
+    appendln("Parallel part interleaving:")
+    append(printInColumnsCustom(interleavingRepresentation) {
+        StringBuilder().apply {
+            for (i in it.indices) {
+                append(if (i == 0) "| " else " | ")
+                append(it[i])
+            }
+            append(" |")
+        }.toString()
+    })
+}
+
+/**
+ * Convert interleaving events to the final form of a matrix of strings
+ */
+private fun splitToColumns(nThreads: Int, interleaving: List<InterleavingEventRepresentation>): List<List<String>> {
+    val result = List(nThreads) { mutableListOf<String>() }
+    for (message in interleaving) {
+        val columnId = message.iThread
+        // write message in an appropriate column
+        result[columnId].add(message.representation)
+        val neededSize = result[columnId].size
+        for (column in result)
+            if (column.size != neededSize)
+                column.add("")
+    }
+    return result
+}
+
+/**
+ * Construct an interleaving graph based on provided [interleavingEvents].
+ * Returns the node corresponding to the first interleaving event.
+ * An interleaving graph consists of two types of edges:
+ * `next` edges form a single-directed list in which the order of events is the same as in [interleavingEvents].
+ * `internalEvents` edges form a directed forest.
+ * The root of each tree is an actor.
+ * Its children are events and method invocations.
+ */
+private fun constructInterleavingGraph(scenario: ExecutionScenario, results: ExecutionResult?, interleavingEvents: List<InterleavingEvent>): InterleavingNode? {
```

Comment:
Why is it called `graph`? It confuses a bit.

Reply:
Because graph theory is a great way to describe complex dependencies.
An interleaving with method calls is not a common object, but a graph is, and it is easier to visualize it as a graph